examples/interfaces/http.py

```diff
@@ -0,0 +1,83 @@
+import json
+from collections import namedtuple
+
+from juju.interface import InterfaceBase
+
+
+class HTTPInterfaceProvides(InterfaceBase):
+    """Provides side for the 'http' interface protocol.
+
+    Example usage:
+
+        class MyCharm(CharmBase):
+            website = Endpoint(HTTPInterfaceProvides)
+
+            def __init__(self, parent, key):
+                super().__init__(parent, key)
+                self.framework.observe(self.website.on.joined, self.on_website_joined)
+
+            def on_website_joined(self, event):
+                self.config = self.framework.model.config
+                for client in self.website.clients:
+                    client.serve(hosts=[client.ingress_address],
+                                 port=self.config['http_port'])
+    """
+    @property
+    def clients(self):
+        return [HTTPInterfaceClient(self.framework, relation) for relation in self.relations]
+
+
+class HTTPInterfaceRequires(InterfaceBase):
+    @property
+    def servers(self):
+        return [HTTPInterfaceServer(relation) for relation in self.relations]
+
+
+class HTTPInterfaceClient:
+    def __init__(self, framework, relation):
+        self._framework = framework
+        self._relation = relation
+
+    @property
+    # TODO: Should this be binding_address?
```

Comment:
We need to expose both and `ingress-address` and `bind-address`.

`bind-address` is the one that will be used for binding a socket to an address. The confusion seems to be about `binding an endpoint to a space` vs `binding a socket to an address`.

More often than not, `ingress-address` is the same as `binding-address` but the logic is complicated by the providers without spaces or if there are no space bindings:
https://github.com/juju/juju/blob/3de9a3afb29daa2e44034d2aef80a12aaa49e3d9/apiserver/facades/agent/uniter/networkinfo.go#L277-L370
https://github.com/juju/juju/blob/3de9a3afb29daa2e44034d2aef80a12aaa49e3d9/apiserver/facades/agent/uniter/uniter.go#L1449-L1471 (EnterScope where this is used to populate relation data).

Retrieving an `ingress-address` for an endpoint binding is odd because an address is associated with a unit, not with a binding or an endpoint.

I see the benefit in having multiple bindings per endpoint though https://discourse.jujucharms.com/t/multiple-space-bindings-per-endpoint/1999/2 so that the same unit may have different {ingress,bind}-address per relation-id.

Reply:
In this case, it's confusing because it's `client.ingress_address` which implies it would be the ingress address sent by the client.  But it should be the address that we should advertise to the client.